lib/index.ts

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2016 Balena
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	 http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/**
+ * @module deviceStatus
+ */
+
+import keyBy = require('lodash/keyBy');
+import maxBy = require('lodash/maxBy');
+
+// This is the earliest possible year since
+// Balena didn't exist before that.
+const RESIN_CREATION_YEAR = 2013;
+
+/**
+ * @summary Map of possible device statuses
+ * @type {Object}
+ * @public
+ * @constant
+ */
+export const status = {
+	CONFIGURING: 'configuring',
+	IDLE: 'idle',
+	OFFLINE: 'offline',
+	INACTIVE: 'inactive',
+	POST_PROVISIONING: 'post-provisioning',
+	UPDATING: 'updating',
+} as const;
+
+export type StatusKey = typeof status[keyof typeof status];
+
+interface Status {
+	key: StatusKey;
+	name: string;
+}
+
+/**
+ * @summary Array of device statuses along with their display names
+ * @type {Object[]}
+ * @public
+ * @constant
+ */
+export const statuses: Status[] = [
+	// The order of statuses in this list is important, as it's reflected
+	// anywhere state are displayed -- currently in the device pie-chart.
+	{ key: status.IDLE, name: 'Online' },
+	{ key: status.CONFIGURING, name: 'Configuring' },
+	{ key: status.UPDATING, name: 'Updating' },
+	{ key: status.OFFLINE, name: 'Offline' },
+	{ key: status.POST_PROVISIONING, name: 'Post Provisioning' },
+	{ key: status.INACTIVE, name: 'Inactive' },
+];
+
+const statusesByKey = keyBy(statuses, 'key');
+
+export interface IHaveProgressAndStatus {
```

Comment:
NIT: Maybe we can group all the interface definitions together.

Reply:
Let me actually move all of them in a different file